accord-core/src/main/java/accord/primitives/AbstractKeys.java

```diff
@@ -229,42 +236,64 @@ public final long foldl(IndexedFoldToLong<? super K> fold, long param, long init
     public final FullKeyRoute toRoute(RoutingKey homeKey)
     {
         if (isEmpty())
-            return new FullKeyRoute(homeKey, new RoutingKey[] { homeKey });
+            return new FullKeyRoute(homeKey, false, new RoutingKey[] { homeKey });
 
-        RoutingKey[] result = toRoutingKeysArray(homeKey);
-        int pos = Arrays.binarySearch(result, homeKey);
-        return new FullKeyRoute(result[pos], result);
+        return toRoutingKeysArray(homeKey, (routingKeys, homeKeyIndex, isParticipatingHomeKey) -> new FullKeyRoute(routingKeys[homeKeyIndex], isParticipatingHomeKey, routingKeys));
     }
 
     protected RoutingKey[] toRoutingKeysArray(RoutingKey withKey)
     {
-        RoutingKey[] result;
-        int resultCount;
-        int insertPos = Arrays.binarySearch(keys, withKey);
-        if (insertPos < 0)
-            insertPos = -1 - insertPos;
+        return toRoutingKeysArray(withKey, (routingKeys, homeKeyIndex, isParticipatingHomeKey) -> routingKeys);
+    }
 
-        if (insertPos < keys.length && keys[insertPos].toUnseekable().equals(withKey))
+    interface ToRoutingKeysFactory<T>
+    {
+        T apply(RoutingKey[] keys, int insertPos, boolean includesKey);
+    }
+
+    @SuppressWarnings("SuspiciousSystemArraycopy")
+    protected <T> T toRoutingKeysArray(RoutingKey withKey, ToRoutingKeysFactory<T> toRoutingKeysFactory)
+    {
+        if (keys.getClass() == RoutingKey[].class)
         {
-            result = new RoutingKey[keys.length];
-            resultCount = copyToRoutingKeys(keys, 0, result, 0, keys.length);
+            int insertPos = Arrays.binarySearch(keys, withKey);
+            if (insertPos >= 0)
+            {
+                Invariants.checkState(keys[insertPos].equals(withKey));
+                return toRoutingKeysFactory.apply((RoutingKey[])keys, insertPos, true);
+            }
+
+            insertPos = -1 - insertPos;
+            RoutingKey[] result = new RoutingKey[1 + keys.length];
+            System.arraycopy(keys, 0, result, 0, insertPos);
+            result[insertPos] = withKey;
+            System.arraycopy(keys, insertPos, result, insertPos + 1, keys.length - insertPos);
+            return toRoutingKeysFactory.apply(result, insertPos, false);
         }
         else
         {
-            result = new RoutingKey[1 + keys.length];
-            resultCount = copyToRoutingKeys(keys, 0, result, 0, insertPos);
-            if (resultCount == 0 || !withKey.equals(result[resultCount - 1]))
-                result[resultCount++] = withKey;
-            resultCount += copyToRoutingKeys(keys, insertPos, result, resultCount, keys.length - insertPos);
+            // TODO review this is a pretty terrible work around for the fact that you can't binarySearch these
```

Comment:
```suggestion
            // TODO (now): review this is a pretty terrible work around for the fact that you can't binarySearch these
```

this project uses a custom TODO syntax that tags comments... I have a script to let me keep track of these tags...

```
$ ./remaming-todos.sh | head
## efficiency: 95
| TODO (required, efficiency): callbacks should be associated with a commandStore for processing to avoid this                  | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| Set<Id> notAwareOfDurability; // TODO (easy, efficiency): use Agrona's IntHashSet as soon as Node.Id switches from long to... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| Set<Id> notPersisted;         // TODO (easy, efficiency): use Agrona's IntHashSet as soon as Node.Id switches from long to... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): we can probably simplify things by requiring (empty) Apply messages to be sent also to the c... | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): forward to local progress shard for processing (if known)                                       | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (desirable, efficiency): if we are co-located with the home shard, don't need to do anything unless we're in a           | modules/accord/accord-core/src/main/java/accord/impl/SimpleProgressLog.java              |
| TODO (find library, efficiency): this is obviously super inefficient, need some range map                                     | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
| TODO (desired, efficiency): this can be made more efficient by batching by epoch                                              | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
| TODO (find lib, efficiency): this is super inefficient, need to store Command in something queryable                          | modules/accord/accord-core/src/main/java/accord/impl/InMemoryCommandStore.java           |
$ ./remaming-todos.sh -m tags
  95 efficiency
  66 low priority
  53 desired
  52 expected
  39 required
  25 testing
  24 consider
  14 easy
  13 now
  13 clarity
   8 desirable
   4 safety
   3 correctness
   3 API
   2 semantics
   2 liveness
   2 exceptions
   2 cleanup
   2 API consistency
   2 ?
   1 validation
   1 urgent
   1 topology
   1 soon
   1 simple
   1 progress
   1 performance
   1 later
   1 find library
   1 find lib
   1 exepected
   1 error handling
   1 behaviour
```

Reply:
I can remove the TODO, Benedict already took a look at it. I'll start using that syntax, thanks.